packages/db-kit/src/utils/fetchExternal.ts

```diff
@@ -0,0 +1,192 @@
+  Fetcher,
+  FetcherConstructor,
+  InvalidNetworkError,
+  default as Fetchers
+} from "@truffle/source-fetcher";
+import type { Project } from "@truffle/db";
+
+export interface FetchExternalOptions {
+  config: TruffleConfig;
+  project: Project.ConnectedProject;
+  address: string;
+}
+
+export async function fetchExternal({
+  config,
+  project,
+  address
+}: FetchExternalOptions): Promise<void> {
+  const { contractName, result } = await fetchAndCompile({ config, address });
+
+  const { contracts } = await project.loadCompile({ result });
+
+  const contract = await findContract({
+    contracts,
+    contractName
+  });
+
+  await project.assignNames({
+    assignments: {
+      contracts: [contract.db.contract]
+    }
+  });
+
+  const { network } = await project.loadMigrate({
+    network: { name: config.network },
+    artifacts: [
+      {
+        ...Common.Shims.NewToLegacy.forContract(contract),
+        networks: {
+          [config.network_id]: {
+            address
+          }
+        }
+      }
+    ]
+  });
+
+  await project.assignNames({
+    assignments: {
+      networks: [network]
+    }
+  });
+}
+
+/**
+ * Accepts a TruffleConfig and returns a priority-sorted list of fetchers based
+ * on provided configuration and whether available fetchers support the current
+ * blockchain network.
+ * @dev HACK this currently only supports EtherscanFetcher
+ */
+export async function forTruffleConfig(options: {
+  config: TruffleConfig;
+}): Promise<Fetcher[]> {
+  const { config } = options;
+  const networkId = parseInt(config.network_id);
+
+  // respect optional `config.sourceFetchers`, used to specify explicit
+  // ordering for which fetchers to use.
+  //
+  // since this HACK-ily only supports Etherscan, this logic serves to respect
+  // configuration that explicitly excludes use of Etherscan.
+  const constructors: FetcherConstructor[] = config.sourceFetchers
+    ? config.sourceFetchers.map(sourceFetcherName => {
+        const Fetcher = Fetchers.find(
+          Fetcher => Fetcher.fetcherName === sourceFetcherName
+        );
+
+        if (!Fetcher) {
+          throw new Error(
+            `Unknown external source servce ${sourceFetcherName}.`
+          );
+        }
+        return Fetcher;
+      })
+    : Fetchers;
+
+  // construct fetchers for networkId
+  const fetchers = await Promise.all(
+    constructors
+      // HACK only support EtherscanFetcher right now
+      .filter(({ fetcherName }) => fetcherName === "etherscan")
+      .map(async (Fetcher: FetcherConstructor) => {
+        const options = config[Fetcher.fetcherName];
+
+        // don't fail on unsupported networks; that's expected behavior
+        try {
+          return await Fetcher.forNetworkId(networkId, options);
+        } catch (error) {
+          if (!(error instanceof InvalidNetworkError)) {
+            throw error;
+          }
+        }
+      })
+  );
+
+  return fetchers.filter((fetcher): fetcher is Fetcher => !!fetcher);
```

Comment:
Sorry, just a syntax question here, since I'm unfamiliar with this syntax.  Is `(x): x is T => e` the same thing as just `(x: T) => e`?  I'm more familiar with the latter, which I think is also clearer/less redundant.

Also as long as I'm nitpicking I will always suggest using `Boolean(x)` rather than `!!x` since converting something to boolean really shouldn't be thought of as double negation... (also alternatives: just `x`, or, in a context like this one, `x !== undefined`...)

Reply:
In this code, I've omitted the argument type because it can be inferred; think of it like this (remember `Fetcher` here is a local variable):
```typescript
(fetcher: BaseFetcher): fetcher is Fetcher => !!fetcher
```

This tells TypeScript that this function will accept any kind of fetcher, and the runtime value will inform the type system that the argument is a _specific_ extension of that fetcher type.

Re: `!!`, this is the style I'm using in @truffle/db, so I'd prefer not to diverge from that here. Good point though, maybe worth changing!

Reply:
Oh -- this is a type predicate.  I see.  I didn't get that; OK, that is pretty different from what I thought it was.  Nevermind.  You should have just said that. :P  (My confusion here is that I thought this was some sort of notation for the argument type, but it's actually the return type.  I'm not used to seeing type predicates as arrow functions...)

OK, nevermind, I get what this is doing now.

Reply:
heh my default is to explain the idea I guess. but I should know that you know terms :)